app/src/cose.h

```diff
@@ -121,92 +220,463 @@ namespace scitt::cose
+        header_items[NOTARY_SIGNING_TIME_INDEX].val.epochDate.nSeconds;
+    }
+    if (
+      header_items[NOTARY_AUTHENTIC_SIGNING_TIME_INDEX].uDataType !=
+      QCBOR_TYPE_NONE)
+    {
+      parsed.notary_authentic_signing_time =
+        header_items[NOTARY_AUTHENTIC_SIGNING_TIME_INDEX]
+          .val.epochDate.nSeconds;
+    }
+    if (header_items[NOTARY_EXPIRY_INDEX].uDataType != QCBOR_TYPE_NONE)
+    {
+      parsed.notary_expiry =
+        header_items[NOTARY_EXPIRY_INDEX].val.epochDate.nSeconds;
+    }
 
     QCBORDecode_ExitMap(&ctx);
     QCBORDecode_ExitBstrWrapped(&ctx);
 
+    qcbor_result = QCBORDecode_GetError(&ctx);
+    if (qcbor_result != QCBOR_SUCCESS)
+    {
+      throw COSEDecodeError(
+        fmt::format("Failed to decode protected header: {}", qcbor_result));
+    }
+
     return parsed;
   }
 
+  UnprotectedHeader decode_unprotected_header(QCBORDecodeContext& ctx)
+  {
+    UnprotectedHeader parsed;
+    // Adapted from parse_cose_header_parameters in t_cose_parameters.c.
+    // t_cose doesn't support custom header parameters yet.
+
+    QCBORError qcbor_result;
+
+    QCBORDecode_EnterMap(&ctx, NULL);
+
+    enum
+    {
+      X5CHAIN_INDEX,
+      END_INDEX,
+    };
+    QCBORItem header_items[END_INDEX + 1];
+    header_items[X5CHAIN_INDEX].label.int64 = COSE_HEADER_PARAM_X5CHAIN;
+    header_items[X5CHAIN_INDEX].uLabelType = QCBOR_TYPE_INT64;
+    header_items[X5CHAIN_INDEX].uDataType = QCBOR_TYPE_ANY;
+    header_items[END_INDEX].uLabelType = QCBOR_TYPE_NONE;
+
+    QCBORDecode_GetItemsInMap(&ctx, header_items);
+
+    qcbor_result = QCBORDecode_GetError(&ctx);
+    if (qcbor_result != QCBOR_SUCCESS)
+    {
+      throw COSEDecodeError(
+        fmt::format("Failed to decode unprotected header: {}", qcbor_result));
+    }
+    if (header_items[X5CHAIN_INDEX].uDataType != QCBOR_TYPE_NONE)
+    {
+      parsed.x5chain = decode_x5chain(ctx, header_items[X5CHAIN_INDEX]);
+    }
+    QCBORDecode_ExitMap(&ctx);
+
+    qcbor_result = QCBORDecode_GetError(&ctx);
+    if (qcbor_result != QCBOR_SUCCESS)
+    {
+      throw COSEDecodeError(
+        fmt::format("Failed to decode unprotected header: {}", qcbor_result));
+    }
+
+    return parsed;
+  }
+
+  std::tuple<ProtectedHeader, UnprotectedHeader> decode_headers(
+    const std::vector<uint8_t>& cose_sign1)
+  {
+    QCBORError qcbor_result;
+
+    QCBORDecodeContext ctx;
+    QCBORDecode_Init(
+      &ctx, cbor::from_bytes(cose_sign1), QCBOR_DECODE_MODE_NORMAL);
+
+    QCBORDecode_EnterArray(&ctx, nullptr);
+    qcbor_result = QCBORDecode_GetError(&ctx);
+    if (qcbor_result != QCBOR_SUCCESS)
+    {
+      throw COSEDecodeError("Failed to parse COSE_Sign1 outer array");
+    }
+
+    uint64_t tag = QCBORDecode_GetNthTagOfLast(&ctx, 0);
+    if (tag != CBOR_TAG_COSE_SIGN1)
+    {
+      throw COSEDecodeError("COSE_Sign1 is not tagged");
+    }
+
+    auto phdr = decode_protected_header(ctx);
+    auto uhdr = decode_unprotected_header(ctx);
```

Comment:
I think it would be good to do a pair of `QCBORDecode_GetByteString`, a `QCBORDecode_CloseArray` and finally a `QCBORDecode_Finish` with an error check here, to ensure the enveloppe is well-formed.

Admittedly t_cose is gonna be doing this eventually, but since this is the first piece of code to be decoding the message it needs to be reasonably robust to garbage.

Reply:
Done in https://github.com/microsoft/scitt-ccf-ledger/pull/90